### ISSUE TITLE
Idea: provide option to output raw LLB

### ISSUE THREAD

**Issue by rrjjvv** [earthly/earthly#2091]
High level, this is pitching the idea that earthly have the capability to output its raw LLB, suitable for piping in to `buildctl debug dump-llb`.

On the surface this might sound silly; one of the things that drew me to earthly was that it gave me (nearly) the full power of pure buildkit without having to learn all the gory details.  Even when on a powerful enough machine where you don't have to throttle the parallelism configs (let alone increase them), it's hard to appreciate how powerful/efficient earthly builds can be when you're looking at a linear stream of interleaved logs.  Seeing a DAG (e.g., the output of `buildctl debug dump-llb --dot`) paints a different picture.

I am doing a presentation to my company about Earthly in my first attempt to actively "sell" it.  I thought a graphical depiction of a build would go a long way (more on that below).  Thus, the motivation for my idea (as stated) is solely as a training/selling tool.  While it _could_ be used as a debugging tool, I doubt it would.  But I probably wouldn't have pitched the idea if training/selling was the only benefit.  I see it as a potential building block towards something *very* valuable: being able to visualize/analyze build hot-spots.  For the obvious/simple example, being able to see the timings (available with `-V`) of targets in a graph itself.  I picture it being analogous to what `pprof` graphs do.

Again, my idea, as stated, is likely of limited use outside of my immediate goal.  I just thought it might make more sense to describe the building block rather than a specific application of that building block.

_(Note: I know it's technically possible; I have a locally hacked client that does exactly what I was proposing.  The primary problem I have at the moment (possibly a result of the way I did my hack) is that earthly "solves" multiple times in one invocation; "merging" the defs for the sake of one graph isn't as straightforward as I thought.  As well, using the actual `buildctl debug dump-llb` command with this hacked output is less than optimal; things like mullti-run `RUN` commands and the earthly-debugger wrappers introduce a lot of undesirable noise.  Removing that noise would probably be out-of-scope for this idea, but I'd imagine that the "merging" of various solves would be in-scope.)_

**Comment by alexcb** [comment]
> The primary problem I have at the moment (possibly a result of the way I did my hack) is that earthly "solves" multiple times in one invocation

For this reason, it's not possible to **only** export the raw LLB, as earthly, in some cases, will block waiting for a reference to be solved, before continuing with generating other separate LLB graphs.

A single pipe redirection would not be possible in this case (instead you would need something like `earthly --llb-exec <some-command>`, which earthly would then execute multiple times.

Given that earthly uses a forked version of buildkit, allowing a custom llb executor would be very fragile.

I like the idea of providing better (any!) visualisation, but in a similar vein, this will be tricky due to earthly blocking in some cases -- which prevents any sort of "dry-run" view of the graph.

**Comment by rrjjvv** [comment]
> For this reason, it's not possible to only export the raw LLB, as earthly, in some cases, will block waiting for a reference to be solved, before continuing with generating other separate LLB graphs.

Yeah, in my first attempt, I was simply dumping each LLB prior to getting sent to the gateway (if I remember the terminology correctly).  I then saw that there was a whole lot of duplication between them (possible a manifestation of the blocking you mention?), so spent time trying to merge them together.  A literal merge didn't have the desired effect, most notably since the vast majority of vertices had multiple edges between them; while _technically_ correct, it's confusing and doesn't reflect what the Earthfile author wrote.  But a more intelligent merge _seems_ to have the desired effect.  The biggest hurdle now is now massaging the exec vertices to be legible (such as stripping out the references to the earthly debugger, dealing with lengthy multi-line `RUN` commands, etc.).  Until I can do that, it's hard to validate whether my merge is truly accurate.

I am *very* new to this; if it sounds like I know what I'm talking about, it's probably beginner's luck 😄 .  But I _think_ the merging I describe above (which is state that has to be maintained/mutated in memory) is another way of describing the problems you mentioned around trying to make earthly behave "just like" buildkit.  I struggled in coming up with a good title; making earthly behave like buildkit (in having earthly be usable in a pipeline, use different executors, "building later", etc.) wasn't my intention at all.  It was truly the visualization I was after.

So if I re-frame the idea as a generic "can earthly generate a build graph/visualization of what it just built", does that make it more feasible?  Ideally it would be a machine-readable format, which could then analyzed, converted into a format suitable for visualization, etc.  The mention of LLB was *solely* because 1) I thought it might make for an easier implementation, which it turn 2) would allow a user to directly make use of `buildctl debug dump-llb`.  Neither is a requirement as far as I'm concerned.  I just thought LLB would result in less reinventing of wheels, but structured data that can be analyzed/visualized is the goal, not the LLB itself.

And conceptually, would my in-memory merging do what I think/hope it'll do?  One comprehensive flow, hiding the fact that it's actually multiple units of work that are somewhat duplicating efforts?  If I'm still overlooking something, then I don't want to waste my time working on something that will just be inaccurate.

**Comment by rrjjvv** [comment]
@alexcb Maybe there is another way to tackle this.  It probably deserves its own standalone issue/request (and I probably will open an issue at some point), but to ease into it: what about using buildkit's native OpenTracing support (https://github.com/moby/buildkit#opentracing-support) as an alternative solution?

The current earthly fork supports this on the server side.  Both native `buildctl` as well as `docker buildx` support this client-side (and works well, even when running against the earthly forked server).  But the earthly client doesn't, and on the server-side, the results from an earthly execution are not exactly friendly (I suspect because earthly's outputs has been geared towards humans, not computers, and the internals have followed suit.)

With this kind of solution, I don't think the fact that there are multiple "solves" would matter (as we're only pushing traces, thus absolved from having to know/remember the relationships or retaining any state).  I'm under the impression that making the build graph a pretty picture would only entail using a custom exporter, and not directly rely on Earthly.
